### PR TITLE
Token wrapper balance and events

### LINKED ISSUE #67: Endpoint for wrapped ETH balance
We need an endpoint for retrieving the balance of wrapped ETH of an user. I am not sure though how the endpoint should be named. I think something like `users/:userAddress/balances/:ethTokenAddress` could be the most consistent regarding our existing endpoints. Or maybe even use `users/:userAddress/balances` and return an object? @berndbohmeier what do you think?

### LINKED ISSUE #71: Endpoint token wrapper events
Add token wrapper contract related events `Deposit`, `Transfer`, `Withdrawal(?)` to new and existing endpoints.

**new endpoints**
`/tokens/:tokenAddress/events`
`/tokens/:tokenAddress/users/:userAddress/events`

**existing endpoints**
`/users/:userAddress/events`

### PR DESCRIPTION
**Pull request by dakingha69**
Closes #67 and closes #71 
- add `/tokens` endpoint
- add `/tokens/:tokenAddress/users/:userAddress/balance` endpoint
- add `/tokens/:tokenAddress/users/:userAddress/events` endpoint
- add events `Deposit`, `Transfer`, `Withdrawal`, `Approval` to existing `/events` endpoint

### CODE REVIEW

File: relay/relay.py [hunk 1]
```diff
@@ -46,11 +49,19 @@ def networks(self) -> Iterable[str]:
     def exchanges(self) -> Iterable[str]:
         return self.orderbook.exchange_addresses
 
+    @property
+    def unw_eth(self) -> Iterable[str]:
```
Comment (dakingha69) [inline relay/relay.py]: change to  smth more comprehensive like `unwrapped_eth`

File: relay/relay.py [hunk 2]
```diff
@@ -110,6 +132,38 @@ def get_networks_of_user(self, user_address: str) -> List[str]:
                 networks_of_user.append(network_address)
         return networks_of_user
 
+    def get_network_event_queries(self, user_address: str, type: str, from_block: int):
```
Comment (berndbohmeier) [inline relay/relay.py]: if you put that function here it should also execute the queries and not only give the queries back.

File: relay/blockchain/unw_eth_proxy.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    DepositEventType,
+    WithdrawalEventType,
+    ApprovalEventType,
+    from_to_types,
+    event_builders
+)
+
+logger = get_logger('unwrap eth', logging.DEBUG)
+
+DepositEvent = 'Deposit'
```
Comment (berndbohmeier) [inline relay/blockchain/unw_eth_proxy.py]: this is already in unw_eth_events.py

Comment (berndbohmeier) [reply relay/blockchain/unw_eth_proxy.py]: remove the duplication

File: relay/blockchain/token_proxy.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    TokenEvent,
+    TransferEventType,
+    ApprovalEventType,
+    from_to_types,
+    event_builders
+)
+
+logger = get_logger('token', logging.DEBUG)
+
+TransferEvent = 'Transfer'
```
Comment (berndbohmeier) [inline relay/blockchain/token_proxy.py]: Duplication, already in token_events.py

File: relay/relay.py [hunk 5]
```diff
@@ -110,6 +132,38 @@ def get_networks_of_user(self, user_address: str) -> List[str]:
...
+                                                 type,
+                                                 user_address=user_address,
+                                                 from_block=from_block))
+            else:
+                queries.append(functools.partial(currency_network_proxy.get_all_network_events,
+                                                 user_address=user_address,
+                                                 from_block=from_block))
+        return queries
+
+    def get_unw_eth_event_queries(self, user_address: str, type: str, from_block: int):
```
Comment (berndbohmeier) [inline relay/relay.py]: same^

File: relay/api/tokens/resources.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
+from relay.blockchain.unw_eth_proxy import UnwEthProxy
+from relay.blockchain.token_proxy import TokenProxy
+
+from relay.api.schemas import TokenEventSchema, UserTokenEventSchema
+from flask_restful import Resource
+from flask import abort
+from webargs import fields
+from webargs.flaskparser import use_args
+from marshmallow import validate
+from typing import Union  # NOQA
```
Comment (berndbohmeier) [inline relay/api/tokens/resources.py]: Please use f code, for example `# noqa: F401` to only disable a specific warning and not all

File: relay/api/tokens/resources.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        abort(404, 'Unkown network: {}'.format(token_address))
+
+
+class TokenAddresses(Resource):
+
+    def __init__(self, trustlines):
+        self.trustlines = trustlines
+
+    def get(self):
+        return self.trustlines.unw_eth_addresses + self.trustlines.token_addresses
```
Comment (berndbohmeier) [inline relay/api/tokens/resources.py]: if you always treat it like this, self.trustlines.token_addresses should just include unw_eth_addresses already, or are there exceptions?

File: relay/api/tokens/resources.py (hunk 6 above)
Comment (dakingha69) [reply relay/api/tokens/resources.py]: done

File: relay/api/tokens/resources.py (hunk 7 above)
Comment (dakingha69) [reply relay/api/tokens/resources.py]: hmm cant think of any exceptions at this moment... will include it.

File: relay/blockchain/token_proxy.py (hunk 4 above)
Comment (dakingha69) [reply relay/blockchain/token_proxy.py]: done

File: relay/blockchain/unw_eth_proxy.py (hunk 3 above)
Comment (dakingha69) [reply relay/blockchain/unw_eth_proxy.py]: done

File: relay/api/tokens/resources.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+from webargs import fields
+from webargs.flaskparser import use_args
+from marshmallow import validate
+from typing import Union  # NOQA
+from relay.relay import TrustlinesRelay
+
+
+def abort_if_unknown_token(trustlines, token_address):
+    if token_address not in trustlines.token_addresses and token_address not in trustlines.unw_eth_addresses:
+        abort(404, 'Unkown network: {}'.format(token_address))
```
[the remark this answers is not in the document]
Comment (dakingha69) [reply relay/api/tokens/resources.py]: fixed

**Comment by berndbohmeier** [comment]
I removed some unused imports. Looks good to me know. To be sure that everything works, we need the e2e tests in the clientlib.

**Pull request merged**